### ISSUE TITLE
Inference file for single image

### ISSUE THREAD

**Issue by Mah-Csiro** [NASA-IMPACT/Prithvi-EO-2.0#16]
Dear @oliverroick, @alukach , @yellowcap , @sunu , and @AliceR 

I have fine-tuned the Prithvi-EO-2.0 model based on the instructions provided in examples/example_multitemporalcrop.ipynb.

Could you please let me know if there is any file or script available to perform inference using the fine-tuned model on a single image?

Kind Regards,
Mah

**Comment by yellowcap** [comment]
Thanks for reaching out @Mah-Csiro , I am not sure if there is an example already for that. In the notebook you mention, the fine tuned head is being trained and used. If you adapt the notebook to use the same dataset, but only use one of the images as input for the classificaition, you would be able to create a fine tuned model that works with only one image.

**Comment by Mah-Csiro** [comment]
Dear @yellowcap,

Thanks for your response. I used the code below to infer one image. Can you please check if the preprocessing and inference method are correct?
I appreciate your time.
import os
import rasterio
import numpy as np
import torch
from einops import rearrange
import albumentations as A
import albumentations
from albumentations import Compose, Flip
from albumentations.pytorch import ToTensorV2
import pandas as pd #import pandas
import lightning.pytorch as pl
from terratorch.tasks import SemanticSegmentationTask
from terratorch.datasets.transforms import FlattenTemporalIntoChannels, UnflattenTemporalFromChannels
from terratorch.datamodules.generic_pixel_wise_data_module import Normalize
import rioxarray #import rioxarray
import matplotlib.pyplot as plt
import matplotlib.colors as mpl_colors
from matplotlib.patches import Rectangle
import numpy as np
import logging
import warnings

class_names = (
        "Natural Vegetation",
        "Forest",
        "Corn",
        "Soybeans",
        "Wetlands",
        "Developed / Barren",
        "Open Water",
        "Winter Wheat",
        "Alfalfa",
        "Fallow / Idle Cropland",
        "Cotton",
        "Sorghum",
        "Other",)

logging.getLogger().setLevel(logging.WARNING)
warnings.filterwarnings("ignore", category=UserWarning, message="DeprecationWarning: scale_modules is deprecated and will be removed in future versions. Use LearnedInterpolateToPyramidal neck instead.")

# --- Configuration ---

CHECKPOINT_PATH = "Prithvi-EO-2/finetuned_model/multicrop_example/checkpoints/best-checkpoint-epoch=48-val_loss=0.00.ckpt"
TEST_IMAGE_PATH = "multi-temporal-crop-classification/validation_chips/chip_002_060_merged.tif"
PREDICTION_OUTPUT_PATH = "Predicted_images/prediction_output.tif"
DATASET_PATH = "multi-temporal-crop-classification" #add the dataset path.
CLASS_WEIGHTS = [0.386375, 0.661126, 0.548184, 0.640482, 0.876862, 0.925186, 3.249462, 1.542289, 2.175141, 2.272419, 3.062762, 3.626097, 1.198702]
LR = 2.0e-4
WEIGHT_DECAY = 0.1
BANDS = ["BLUE", "GREEN", "RED", "NIR_NARROW", "SWIR_1", "SWIR_2"]
NUM_FRAMES = 3
MEANS = {"BLUE": 494.905781, "GREEN": 815.239594, "RED": 924.335066, "NIR_NARROW": 2968.881459, "SWIR_1": 2634.621962, "SWIR_2": 1739.579917}
STDS = {"BLUE": 284.925432, "GREEN": 357.84876, "RED": 575.566823, "NIR_NARROW": 896.601013, "SWIR_1": 951.900334, "SWIR_2": 921.407808}

# --- Load Model ---
HEAD_DROPOUT = 0.1
FREEZE_BACKBONE = False
backbone_args = dict(backbone_pretrained=True, backbone="prithvi_eo_v2_600_tl", backbone_coords_encoding=["time", "location"], backbone_bands=BANDS, backbone_num_frames=NUM_FRAMES)
decoder_args = dict(decoder="UperNetDecoder", decoder_channels=256, decoder_scale_modules=True)
necks = [dict(name="SelectIndices", indices=[7, 15, 23, 31]), dict(name="ReshapeTokensToImage", effective_time_dim=NUM_FRAMES)]
model_args = dict(**backbone_args, **decoder_args, num_classes=len(CLASS_WEIGHTS), head_dropout=HEAD_DROPOUT, necks=necks, rescale=True)
model = SemanticSegmentationTask.load_from_checkpoint(checkpoint_path=CHECKPOINT_PATH, model_args=model_args, class_weights=CLASS_WEIGHTS, loss="ce", lr=LR, optimizer="AdamW", optimizer_hparams=dict(weight_decay=WEIGHT_DECAY), ignore_index=-1, freeze_backbone=FREEZE_BACKBONE, freeze_decoder=False, model_factory="EncoderDecoderFactory")
model.eval()
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device)

# --- Load Image and Metadata ---
with rasterio.open(TEST_IMAGE_PATH) as src:
    data = src.read()
    profile = src.profile
data[data == -9999] = 0

metadata_df = pd.read_csv(os.path.join(DATASET_PATH,"chips_df.csv"))
image_name = os.path.basename(TEST_IMAGE_PATH).replace("_merged.tif","")
row = metadata_df[metadata_df['chip_id'] == image_name].iloc[0]

# --- Preprocessing ---
reshaped_image = rearrange(data, "(time channels) height width -> time height width channels", time=NUM_FRAMES, channels=6)
transform = Compose([FlattenTemporalIntoChannels(),  ToTensorV2(), UnflattenTemporalFromChannels(n_timesteps=NUM_FRAMES)])
transformed = transform(image=reshaped_image)
image_tensor = transformed['image'].unsqueeze(0).float().to(device)

normalize_transform = Normalize(means=list(MEANS.values()), stds=list(STDS.values()))
image_tensor = normalize_transform({"image": image_tensor})["image"]


# --- Metadata Tensors ---
def _get_date(row):
    temporal_coords = []
    date_columns = ["first_img_date", "middle_img_date", "last_img_date"]
    for col in date_columns:
        date_str = row[col]
        date = pd.to_datetime(date_str, format="%Y-%m-%d")
        temporal_coords.append([date.year, date.dayofyear - 1])
    return torch.tensor(temporal_coords, dtype=torch.float32).unsqueeze(0).to(device)

def _get_coords(image_path):
    image = rioxarray.open_rasterio(image_path, masked=True)
    px = image.x.shape[0] // 2
    py = image.y.shape[0] // 2
    point = image.isel(band=0, x=slice(px, px + 1), y=slice(py, py + 1))
    point = point.rio.reproject("epsg:4326")
    lat_lon = np.asarray([point.y[0], point.x[0]])
    return torch.tensor(lat_lon, dtype=torch.float32).unsqueeze(0).to(device)

temporal_coords = _get_date(row)
location_coords = _get_coords(TEST_IMAGE_PATH)

print(f'temporal_coords is {temporal_coords}')
print(f'location_coords is {location_coords}')
# --- Prediction ---
with torch.no_grad():
    prediction = model(image_tensor, temporal_coords=temporal_coords, location_coords=location_coords)
    prediction = torch.argmax(prediction.output, dim=1).squeeze(0).cpu().numpy()
    print(prediction)


   
    
# --- Save Prediction ---
profile.update(
    dtype=rasterio.uint8,
    nodata=None,
    count=1,
    blockysize=36, # correct the block size
    blockxsize=224, # correct the block size
    compress='lzw'
)

with rasterio.open(PREDICTION_OUTPUT_PATH, 'w', **profile) as dst:
    dst.write(prediction.astype(rasterio.uint8), 1)

print(f"Prediction saved to {PREDICTION_OUTPUT_PATH}")

# --- Show Prediction (Optional) ---

def plot_prediction(prediction_mask, class_names):
    """
    Plots the prediction mask with a colorbar and legend.

    Args:
        prediction_mask (numpy.ndarray): The predicted segmentation mask.
        class_names (list): List of class names.
    """
    plt.figure(figsize=(12, 6))
    plt.imshow(prediction_mask, cmap='jet')
    plt.title("Predicted Segmentation Mask")
    plt.colorbar()

    num_classes = len(class_names)
    norm = mpl_colors.Normalize(vmin=0, vmax=num_classes - 1)
    cmap = plt.get_cmap("jet")

    legend_data = [[i, cmap(norm(i)), class_names[i]] for i in range(num_classes)]
    handles = [Rectangle((0, 0), 1, 1, color=tuple(v for v in c)) for k, c, n in legend_data]
    labels = [n for k, c, n in legend_data]
    plt.legend(handles, labels, loc="center", bbox_to_anchor=(0.5, -0.15), ncol=3) #adjust legend location as needed.
    plt.tight_layout()
    plt.show()
    
plot_prediction(prediction,class_names)

**Comment by chs412** [comment]
How can I perform inference on single time-step image with 6 hls bands only? 
I have trained the model on 3 time steps (16 bands) for classfication task.
